pkg/extensions/sync/utils.go

```diff
@@ -89,6 +100,284 @@ func getFileCredentials(filepath string) (CredentialsFile, error) {
+
+	mResp, err := client.R().
+		Get(getCosignManifestURL.String())
+	if err != nil {
+		log.Error().Err(err).Str("url", getCosignManifestURL.String()).
+			Msgf("couldn't get cosign manifest: %s", cosignEncodedDigest)
+		return err
+	}
+
+	if mResp.IsError() {
+		log.Info().Msgf("couldn't find any cosign signature from %s, status code: %d skipping",
+			getCosignManifestURL.String(), mResp.StatusCode())
+		return nil
+	}
+
+	var m ispec.Manifest
+
+	err = json.Unmarshal(mResp.Body(), &m)
+	if err != nil {
+		log.Error().Err(err).Str("url", getCosignManifestURL.String()).
+			Msgf("couldn't unmarshal cosign manifest %s", cosignEncodedDigest)
+		return err
+	}
+
+	imageStore := storeController.GetImageStore(repo)
+
+	for _, blob := range m.Layers {
+		// get blob
+		getBlobURL := regURL
+		getBlobURL.Path = path.Join(getBlobURL.Path, "v2", repo, "blobs", blob.Digest.String())
+
+		resp, err := client.R().SetDoNotParseResponse(true).
+			Get(getBlobURL.String())
+		if err != nil {
+			log.Error().Err(err).Msgf("couldn't get cosign blob: %s", blob.Digest.String())
+			return err
+		}
+
+		if resp.IsError() {
+			log.Info().Msgf("couldn't find cosign blob from %s, status code: %d", getBlobURL.String(), resp.StatusCode())
+			return errors.ErrBadBlobDigest
+		}
+
+		defer resp.RawBody().Close()
+
+		// push blob
+		_, _, err = imageStore.FullBlobUpload(repo, resp.RawBody(), blob.Digest.String())
+		if err != nil {
+			log.Error().Err(err).Msg("couldn't upload cosign blob")
+			return err
+		}
+	}
+
+	// get config blob
+	getBlobURL := regURL
+	getBlobURL.Path = path.Join(getBlobURL.Path, "v2", repo, "blobs", m.Config.Digest.String())
+
+	resp, err := client.R().SetDoNotParseResponse(true).
+		Get(getBlobURL.String())
+	if err != nil {
+		log.Error().Err(err).Msgf("couldn't get cosign config blob: %s", getBlobURL.String())
+		return err
+	}
+
+	if resp.IsError() {
+		log.Info().Msgf("couldn't find cosign config blob from %s, status code: %d", getBlobURL.String(), resp.StatusCode())
+		return errors.ErrBadBlobDigest
+	}
+
+	defer resp.RawBody().Close()
+
+	// push config blob
+	_, _, err = imageStore.FullBlobUpload(repo, resp.RawBody(), m.Config.Digest.String())
+	if err != nil {
+		log.Error().Err(err).Msg("couldn't upload cosign blob")
+		return err
+	}
+
+	// push manifest
+	_, err = imageStore.PutImageManifest(repo, cosignEncodedDigest, ispec.MediaTypeImageManifest, mResp.Body())
+	if err != nil {
+		log.Error().Err(err).Msg("couldn't upload cosing manifest")
+		return err
+	}
+
+	return nil
+}
+
+func syncNotarySignature(client *resty.Client, storeController storage.StoreController,
+	regURL url.URL, repo, digest string, log log.Logger) error {
+	log.Info().Msg("syncing notary signatures")
+
+	getReferrersURL := regURL
+	getRefManifestURL := regURL
+
+	// based on manifest digest get referrers
+	getReferrersURL.Path = path.Join(getReferrersURL.Path, "oras/artifacts/v1/", repo, "manifests", digest, "referrers")
```

Comment:
same, does notation lib have helper libs?

Reply:
no, they are doing the same